fcm_django/models.py

```diff
@@ -387,3 +388,5 @@ class Meta:
         indexes = [
             models.Index(fields=["registration_id", "user"]),
         ]
+
+        swappable = swapper.swappable_setting("fcm_django", "fcmdevice")
```

Comment:
I also believe that inheriting from `AbstractFCMDevice` should be mandatory in case one wants to use a custom Device model for us to ensure that logic is preserved. But in that case we should completely refactor our models. We probably should move all the fields from abstract models to `FCMDevice` model, and in abstract models we can use class attributes corresponding to fields that are used in business logic, similar to what Django does in User model. For our case this can look like this:

```python
class AbstractFCMDevice:
    TOKEN_FIELD: str

    @property
    def token(self):
        return getattr(self, self.TOKEN_FIELD)

    def send_message(self, message, app, **kwargs):
        message.token = self.token
        ...


class FCMDevice(AbstractFCMDevice):
    registration_id = models.CharField()

    TOKEN_FIELD = "registration_id"
```

Reply:
I also think inheriting from AbstractFCMDevice should be mandatory. Otherwise the entire purpose of this package is void and it breaks logic

Reply:
> I also think inheriting from AbstractFCMDevice should be mandatory. Otherwise the entire purpose of this package is void and it breaks logic

What should be done for that? Mention in the doc that _CustomDevice_ should be always inherited from `AbstractFCMDevice`?

Reply:
At the very least that yeah 🤔

Reply:
I would remove that line, yep

Reply:
> I would remove that line, yep

I have changed those lines in the docs.

Reply:
There is one more thing from this conversation to pull out -> we should skip migrating FCMDevice, if the user has decided to overwrite the model. There is no reason for the user to have two tables, one that they do not use

Reply:
> There is one more thing from this conversation to pull out -> we should skip migrating FCMDevice, if the user has decided to overwrite the model. There is no reason for the user to have two tables, one that they do not use

In order to accomplish that user don't have to register `fcm_django` at `INSTALLED_APPS`. I updated docs and settings for tests.

Reply:
I'm not 100% sure about this solution to be honest. This will be problematic if we ever add any other models. Not saying we will, but you know.. What do you think @olzhasar

Reply:
> This will be problematic if we ever add any other models. Not saying we will, but you know.

@xtrinch 
If there will be more models and few of them is required it could be done in the same way as it done at [allauth](https://django-allauth.readthedocs.io/en/latest/installation.html). Basically there are an app with nested apps. So to use fcm-django with integrated Device model user will have to add into installed apps:
```python
"fcm_django",
"fcm_django.device",
```

If user would like to use only required models out of `fcm_django` then adds only 
```python
"fcm_django"
```

And these changes can be made in the future, at this stage no modifications are required as there is only one model in the application.

Reply:
@xtrinch I do agree that migrations for our default device model will become obsolete if one decides to use a custom model, but it's probably not that critical. There are other a lot more concerning issues that this PR entails.

Reply:
@olzhasar Yeah, especially if it can be skipped by skipping it in installed_apps. Is there anything besides the comments you already left?

Reply:
@xtrinch My biggest concern is still about introducing external dependency. I did not receive any valid reasons for doing that. I've just created a sample pull request to demonstrate how can we implement `swappable` behavior without using `swapper`. Here is a one particular commit you should see:
https://github.com/xtrinch/fcm-django/pull/248/commits/66039eaef7c30a2de6950ae42653acf72529950f

Reply:
I also think that we should redesign abstract models to prepare everything for this PR. One of the most common reasons why one might be interested in using a custom Device model is because they don't need some fields that standard model provides. I don't see how the current implementation helps with that. Let's say I don't want `device_id` in my model. This field is not involved in any business logic and it's a reasonable wish. However, the only way for me to exclude it from my table (correct me if I'm wrong) is to set it to `None` in my custom model, e.g.:

```py3
class CustomDevice(AbstractFCMDevice):
    `device_id` = None
```

This way of doing feels completely unnatural. As a user, I would instead like `fcm-django` to keep only necessary fields in the abstract model and allow me to handle the rest.

Another concern is the admin interface. If I want to customize it, I will probably need to deregister the old admin and register a new one. This feels odd as well.

Reply:
> I've just created a sample pull request to demonstrate how can we implement `swappable` behavior without using `swapper`. Here is a one particular commit you should see

Your code 60 lines without doc strings and tests. Swapper contain 100 lines of code with doc strings and also contains tests. 

> One of the most common reasons why one might be interested in using a custom Device model is because they don't need some fields that standard model provides.

But in this PR it was your concern that inheritance from AbstractDevice should be mandatory. If there's no inheritance and user would like to have fully customized model then there's no problem.

> Another concern is the admin interface. If I want to customize it, I will probably need to deregister the old admin and register a new one. This feels odd as well.

Same as with customization User at Django itself.

Reply:
I have received many requests over the years to be able to add fields to FCMDevice, never to remove them. I think removing fields is actually an edge case that may break something if the wrong field is chosen

Reply:
@Akay7 Which 60 lines are you referring to? The replacement of `swapper` in my code is just a single function that is 14 lines of code. We can add docstrings and tests to this function prior to merging. The code is just to show you the idea that we don't any external library for our simple use case

Reply:
@xtrinch Yes, removing fields should not be allowed, that's the reason why I think that abstract model should contain only necessary fields.
I used `fcm-django` in one project of mine and I didn't need `device_id` field there. So, not being forced to add it to my Custom Device model would have been great.

Also, I wonder what is the motivation for having two abstract models: `Device` and `FCMDevice`?

Reply:
> Which 60 lines are you referring to? The replacement of `swapper` in my code is just a single function that is 14 lines of code.

@olzhasar okay, my bad it's 14 lines. 
And there will be 14 lines for two models or it will be 28? Yeah, it's possible to implement swapper on our own but after reaching a certain complexity of the project, it will look about the same as the swapper. But that would also take up our time.